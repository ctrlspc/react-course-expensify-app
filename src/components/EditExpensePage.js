import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component {
  onExpenseFormSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemoveClick = (id) => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/');
  }

  render () {
    return (
      <div>
        <ExpenseForm 
          onSubmit={this.onExpenseFormSubmit}
          expense={this.props.expense} />
          <button onClick={this.onRemoveClick}>
            Remove
          </button>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense({id}))
});

const mapStateToProps = (state, props) => {  
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};


export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);