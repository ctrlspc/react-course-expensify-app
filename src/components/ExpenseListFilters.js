import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate , setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocussed:null
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocussed) => this.setState({ calendarFocussed });
  
  onFilterTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onFilterSelectChange = (e) => {
    if(e.target.value === 'date'){
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    const props = this.props;
    return (
    <div>
      <input 
        //this is a controlled input - becuase we are programatically controlling the value and onchange
        type="text" 
        value={this.props.filters.text} 
        onChange={this.onFilterTextChange}
      />
      <select 
        //this is also a controlled input
        value={this.props.filters.sortBy} 
        onChange={this.onFilterSelectChange}
      >
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
      <DateRangePicker
        startDate={this.props.filters.startDate} 
        endDate={this.props.filters.endDate} 
        onDatesChange={this.onDatesChange} 
        focusedInput={this.state.calendarFocussed} 
        onFocusChange={this.onFocusChange} 
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
    </div>
    )
  }
}

const mapStateToProps = (state) => ({
  filters:state.filters
})

const mapDispatchToProps = (dispatch) => ({
  sortByAmount: () => dipatch(sortByAmount()) ,
  sortByDate: () => dispatch(sortByDate()) ,
  setTextFilter: (text) => dipatch(setTextFilter(text)) ,
  setEndDate:  (endDate) => dipatch(setEndDate(endDate)),
  setStartDate: (startDate) => dispatch(setStartDate(startDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)