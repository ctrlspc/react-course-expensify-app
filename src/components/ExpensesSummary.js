import React from 'react'
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import 'numeral/locales';
import numeral from 'numeral';

numeral.locale('en-gb');

export const ExpensesSummary = (props) => {
  
  const expenseWord = props.expensesCount > 1 ? 'expenses' : 'expense';
  const formattedTotal = numeral(props.expensesTotal/100).format('$0,0.00');
  return (<div>
    Viewing {props.expensesCount} {expenseWord} totalling {formattedTotal}
  </div>
  );
};

const mapStateToProps = (state) => {
  const filteredExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expensesCount:filteredExpenses.length,
    expensesTotal:selectExpensesTotal(filteredExpenses)
  }
};



export default connect(mapStateToProps)(ExpensesSummary);

