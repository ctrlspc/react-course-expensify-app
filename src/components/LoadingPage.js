import React from 'react';

const LoadingPage = () => (
  <div className="loader">
    <img className="loader__image" src="/images/loader.gif" alt="spinngin loader gif"/>
  </div>
)

export default LoadingPage;