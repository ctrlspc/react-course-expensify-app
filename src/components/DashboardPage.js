import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpensesSummary';

const DashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseSummary />
    <ExpenseList />
  </div>
  
);

export default DashboardPage;