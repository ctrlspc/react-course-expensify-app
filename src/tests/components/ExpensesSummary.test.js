import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary'
import { shallow } from 'enzyme';

test('should correctly render ExpensesSummary component for single expense' , () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={1000} />);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary component for multiple expenses' , () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={1000} />);
  expect(wrapper).toMatchSnapshot();
});