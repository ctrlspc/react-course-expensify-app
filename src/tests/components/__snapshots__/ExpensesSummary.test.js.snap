// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary component for multiple expenses 1`] = `
<div>
  Viewing 
  2
   
  expenses
   totalling 
  £10.00
</div>
`;

exports[`should correctly render ExpensesSummary component for single expense 1`] = `
<div>
  Viewing 
  1
   
  expense
   totalling 
  £10.00
</div>
`;
